### PR TITLE
Optimized SparkRunner ParDo Operation

### LINKED ISSUE #32537: [Task]: Optimize Spark Runner parDo transform evaluator
### What needs to happen?

When evaluating ParDo operations in the TransformTranslator in Apache Spark Runner, too many filter operations are applied.
The reason for applying filter operations is that a ParDo can have multiple outputs, so we apply filter operations to filter only elements such as each TupleTag.

However, the filter operation is also applied to a ParDo with a single output, which can have a performance impact.
Therefore, we should avoid applying the filter operation when evaluating ParDo operations with a single output.

[related mail context](https://lists.apache.org/thread/oj5ntt2xrxo74yl47953gm2b9vmwgd8f)

### Issue Priority

Priority: 2 (default / most normal work should be filed as P2)

### Issue Components

- [ ] Component: Python SDK
- [X] Component: Java SDK
- [ ] Component: Go SDK
- [ ] Component: Typescript SDK
- [ ] Component: IO connector
- [ ] Component: Beam YAML
- [ ] Component: Beam examples
- [ ] Component: Beam playground
- [ ] Component: Beam katas
- [ ] Component: Website
- [ ] Component: Infrastructure
- [X] Component: Spark Runner
- [ ] Component: Flink Runner
- [ ] Component: Samza Runner
- [ ] Component: Twister2 Runner
- [ ] Component: Hazelcast Jet Runner
- [ ] Component: Google Cloud Dataflow Runner

### PR DESCRIPTION
**Pull request by twosom**
**Please** add a meaningful description for your change here

fixes #32537 

This PR contains these changes
- optimized to skip filter operation when there is only single output
- add related test

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://github.com/apache/beam/blob/master/CONTRIBUTING.md#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI or the [workflows README](https://github.com/apache/beam/blob/master/.github/workflows/README.md) to see a list of phrases to trigger workflows.

### CODE REVIEW

**Comment by Abacn** [comment]
Thanks for the change. I saw several changes of Spark runner improvement this PR and #32610 . If you are looking for reviewer a suggestion here is to send an email to beam devlist (https://lists.apache.org/list.html?dev@beam.apache.org) introducing these changes and ask people who are familiar with Beam Spark Runner for review

File: runners/spark/src/main/java/org/apache/beam/runners/spark/translation/TransformTranslator.java [hunk 1]
```diff
@@ -463,17 +464,28 @@ public void evaluate(
...
-              (JavaRDD<WindowedValue<Object>>) (JavaRDD<?>) filtered.values();
-          context.putDataset(output.getValue(), new BoundedDataset<>(values));
+              (JavaRDD<WindowedValue<Object>>) (JavaRDD<?>) all.values();
+          context.putDataset(
+              Iterables.getOnlyElement(outputs.entrySet()).getValue(),
+              new BoundedDataset<>(values));
         }
       }
 
+      private boolean hasMultipleOutputs(Map<TupleTag<?>, PCollection<?>> outputs) {
```
Comment (Abacn) [inline runners/spark/src/main/java/org/apache/beam/runners/spark/translation/TransformTranslator.java:485]: this can be static

**Review by Abacn — APPROVED** [review]
thanks, just had a minor comment

By any chance is it able to test the performance diff?

File: runners/spark/src/main/java/org/apache/beam/runners/spark/translation/TransformTranslator.java (hunk 1 above)
Comment (Abacn) [inline runners/spark/src/main/java/org/apache/beam/runners/spark/translation/TransformTranslator.java:485]:
```suggestion
      private static boolean hasMultipleOutputs(Map<TupleTag<?>, PCollection<?>> outputs) {
```

File: runners/spark/src/main/java/org/apache/beam/runners/spark/translation/TransformTranslator.java [hunk 2]
```diff
@@ -463,17 +464,28 @@ public void evaluate(
...
-              (JavaRDD<WindowedValue<Object>>) (JavaRDD<?>) filtered.values();
-          context.putDataset(output.getValue(), new BoundedDataset<>(values));
+              (JavaRDD<WindowedValue<Object>>) (JavaRDD<?>) all.values();
+          context.putDataset(
+              Iterables.getOnlyElement(outputs.entrySet()).getValue(),
+              new BoundedDataset<>(values));
         }
       }
 
+      private static boolean hasMultipleOutputs(Map<TupleTag<?>, PCollection<?>> outputs) {
```
Comment (Abacn) [inline runners/spark/src/main/java/org/apache/beam/runners/spark/translation/TransformTranslator.java:485]: ah sorry, this is inside an anonymous class

Comment (Abacn) [reply runners/spark/src/main/java/org/apache/beam/runners/spark/translation/TransformTranslator.java:485]:
```suggestion
      private boolean hasMultipleOutputs(Map<TupleTag<?>, PCollection<?>> outputs) {
```

**Comment by twosom** [comment]
> thanks, just had a minor comment
> 
> By any chance is it able to test the performance diff?

@Abacn 
I tested it on a 40GB DNA dataset.
It's a pipeline that simply counts the number of DNA elements.
Firstly, the DAG is less verbose because it's stripped of unnecessary filters.
In terms of performance, unfortunately, I didn't see a dramatic change at 40GB, which I don't think you'll notice unless you're working with very large datasets, as it simply removes redundant filter operations.
(It did have a slight effect though).
In the picture, the left side is the pipeline after this PR, and the right side is the old pipeline.



image

image

**Pull request merged**